packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1991,8 +2598,241 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+      // pageless routes do not need to be skipped.
+      pagelessRoutesToSkip = 0;
+      oldEntriesTop -= 1;
+      newPagesTop -= 1;
+    }
+    // Reverts the pageless routes that cannot be updated.
+    oldEntriesTop += pagelessRoutesToSkip;
+
+    // Scans middle of the old entries and records the page key to old entry map.
+    int oldEntriesBottomToScan = oldEntriesBottom;
+    final Map<LocalKey, _RouteEntry> pageKeyToOldEntry = <LocalKey, _RouteEntry>{};
+    while (oldEntriesBottomToScan <= oldEntriesTop) {
+      final _RouteEntry oldEntry = _history[oldEntriesBottomToScan];
+      oldEntriesBottomToScan += 1;
+      assert(
+        oldEntry != null &&
+        oldEntry.currentState != _RouteLifecycle.disposed
+      );
+      // Pageless routes will be recorded when we update the middle of the old
+      // list.
+      if (!oldEntry.hasPage)
+        continue;
+
+      assert(oldEntry.hasPage);
+
+      final Page<dynamic> page = oldEntry.route.settings as Page<dynamic>;
+      if (page.key == null)
+        continue;
+
+      assert(!pageKeyToOldEntry.containsKey(page.key));
+      pageKeyToOldEntry[page.key] = oldEntry;
+    }
+
+    // Scans middle of the new pages and records page key to new page map.
+    int newPagesBottomToScan = newPagesBottom;
+    final Map<LocalKey, Page<dynamic>> pageKeyToNewPage = <LocalKey, Page<dynamic>>{};
+    while (newPagesBottomToScan <= newPagesTop) {
+      final Page<dynamic> page = widget.pages[newPagesBottomToScan];
+      if (page.key != null) {
+        assert(!pageKeyToNewPage.containsKey(page.key), 'Duplicated page key.');
+        pageKeyToNewPage[page.key] = page;
+      }
+      newPagesBottomToScan += 1;
+    }
+
+    // Updates the middle of the list.
+    while (newPagesBottom <= newPagesTop) {
+      final Page<dynamic> nextPage = widget.pages[newPagesBottom];
+      newPagesBottom += 1;
+      if (
+        nextPage.key == null ||
+        !pageKeyToOldEntry.containsKey(nextPage.key) ||
+        !pageKeyToOldEntry[nextPage.key].canUpdateFrom(nextPage)
+      ) {
+        // There is no matching key in the old history, we need to create a new
+        // route and wait for the transition delegate to decide how to add
+        // it into the history.
+        final _RouteEntry newEntry = _RouteEntry(
+          nextPage.createRoute(context),
+          initialState: _RouteLifecycle.staging,
+        );
+        assert(
+          newEntry.route.settings == nextPage,
+          'If a route is created from a page, its must have that page as its '
+          'settings.',
+        );
+        newHistory.add(newEntry);
+      } else {
+        // Removes the key from pageKeyToOldEntry to indicate it is taken.
+        final _RouteEntry matchingEntry = pageKeyToOldEntry.remove(nextPage.key);
+        assert(matchingEntry.canUpdateFrom(nextPage));
+        matchingEntry.route._updateSettings(nextPage);
+        newHistory.add(matchingEntry);
```

Comment:
What about the pageless routes above matchingEntry?

Reply:
the new history only contains page-based route all the pageless route will be inserted back after transition delegate resolves them